### DISCUSSION TITLE
[@types/hapi__hapi] Catch variable is incompatible with log signature

### DISCUSSION THREAD

**Discussion by iybaron** [DefinitelyTyped/DefinitelyTyped#55752] [Issues with a @types package] [answered]
In Typescript v4.4, a new flag, `--useUnknownInCatchVariables`, was added to change the default type of catch clause variables from `any` to `unknown`. This flag is enabled under the `--strict` family of options. https://devblogs.microsoft.com/typescript/announcing-typescript-4-4-rc/#use-unknown-catch-variables

Variables of type `unknown` are incompatible with the Hapi `server.log` and `request.log` function signatures. Therefore, Typescript does not allow this:
```
try {
  // do something
} catch (error) {
  request.log(tags, error)
}
```
On the call to `request.log`, the TS compiler outputs `Type 'unknown' is not assignable to type '() => string | object'`.

NOTE: It is possible to get around this issue by adding type guards or sanitizing the catch variable.

[a comment by an automated account is not shown]

**Comment by saboya** [comment, marked as answer]
Well, you turned on a feature that is more strict regarding type safety. Adding type guards is not a "workaround", it is literally what you opted into doing when you enabled that flag.

**Reply by iybaron** [reply]
That's fair enough. The flip-side of that token is that the data type on Request.log might be unnecessarily restrictive, so coercing the `unknown` into something loggable becomes unintuitive (or _potentially_ inaccurate). For example, is it true that the Hapi logger can't handle numbers or null?

**Reply by saboya** [reply]
Typings are usually written based on the official API documentation (whenever possible / available). [This is what it says](https://hapi.dev/api/?v=20.1.4#-serverlogtags-data-timestamp):

> `data` - (optional) an message string or object with the application data being logged. If `data` is a function, the function signature is `function()` and it called once to generate (return value) the actual data emitted to the listeners. If no listeners match the event, the `data` function is not invoked.

The [actual code](https://github.com/hapijs/hapi/blob/ab4039492782096e5f9d772ffeed3de2481ccff6/lib/core.js#L628) suggests that it also checks for `Error` instances.

The [current Typescript definition is](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c4ef58658f938611a8596a8df77d4769613ac5f5/types/hapi__hapi/index.d.ts#L594):
```Typescript
log(tags: string | string[], data?: string | object | (() => string | object)): void;
```

Now, does the actual code handles `number` or `null`? According to the code, yes, it should handle those. However, given that the official API does not say that, I'd argue that by leaving it out the unofficial typings are doing a better job at enforcing the correct use of the API.

You could make a case to add `null` there and in my opinion I wouldn't say it's wrong, but accepting `unknown` is 100% wrong in my opinion. The whole idea behind that flag is to actually enforce type checks by making the default catch type `unknown` instead of `any`. If that doesn't feel ergonomic for you as a developer, I'd tell you to drop the flag.

**Reply by iybaron** [reply]
Thanks @saboya.
